test/src/test/java/hudson/model/JobTest.java

```diff
@@ -299,6 +311,107 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
         p.renameTo("different-name");
     }
 
+    @Issue("JENKINS-44657")
+    @Test public void testRenameWithCustomBuildsDirWithBuildsIntact() throws Exception {
+        j.jenkins.setRawBuildsDir("${JENKINS_HOME}/builds/${ITEM_FULL_NAME}/builds");
+        final FreeStyleProject p = j.createFreeStyleProject();
+        final File oldBuildsDir = p.getBuildDir();
+        j.buildAndAssertSuccess(p);
+        String oldDirContent = dirContent(oldBuildsDir);
+        p.renameTo("different-name");
+        final File newBuildDir = p.getBuildDir();
+        assertNotNull(newBuildDir);
+        assertNotEquals(oldBuildsDir.getAbsolutePath(), newBuildDir.getAbsolutePath());
+        String newDirContent = dirContent(newBuildDir);
+        assertEquals(oldDirContent, newDirContent);
+    }
+
+    @Issue("JENKINS-44657")
+    @Test public void testRenameWithCustomBuildsDirWithBuildsIntactInFolder() throws Exception {
+        j.jenkins.setRawBuildsDir("${JENKINS_HOME}/builds/${ITEM_FULL_NAME}/builds");
+        final Folder f = j.jenkins.createProject(Folder.class, "F");
```

Comment:
Generally the only thing `MockFolder` lacks is a facet allowing `WebClient`-based tests to browse its index page (browsing jobs contained inside it works fine).